text/0000-shared-layers.md

```diff
@@ -0,0 +1,117 @@
+- build layers: Layers from a specific buildpack that are available to subsequent buildpacks during the build process.
+
+# Motivation
+[motivation]: #motivation
+
+## Why should we do this?
+
+Currently the spec forbids layer modifictions by buildpacks that don't contribute a specific layer. This is not enforced by the lifecycle and leads to cases where subsequent buildpacks can unknowningly modify layer content as a side-effect of executing binaries provided by a build layer. Such unexpected modifications are also really hard to debug and pin-point the case as it can be really difficult to figure out which commands from subsequent buildpacks are causing them. An RFC was created to make build layers read-only for subsequent buildpacks in [PR#155](https://github.com/buildpacks/rfcs/pull/155)
+
+On the other hand there are valid use cases for build layers that can be used as a collaborative workspace for multiple buildpacks and need to be cached in subsequent runs. This RFC is meant to provide a solution for such situations.
+
+
+## What use cases does it support?
+
+There are various use cases when you may want a common workspace that multiple buildpacks can collaborate on during the build process and this workspace needs to be cached, hence it can't be a temporary directory.
+
+In such a case a buildpack could create a "shared layer" i.e. subsequent buildpacks can modify its contents and the final state of the layer at the end of the build process is what is exported out.
+
+One example can be a CCache buildpack that sets up CCACHE_DIR environment variable (pointing to a directory in a shared layer), which a CMake buildpack may use for its build cache.
+
+Another example can be a pip cache buildpack that sets up a PIP_CACHE_DIR environment variable (pointing to a directory in a shared layer), which could be used by a [pip-tools](https://github.com/jazzband/pip-tools) buildpack, or a [pipenv](https://pipenv.pypa.io/en/latest/) buildpack to resolve the python dependency tree and generate a requirements.txt and subsequently the pip buildpack could install these requirement using the wheels cached during the dependency resolution.
+
+- What is the expected outcome?
+
+The spec and lifecycle is modified to support the above use cases.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC introduces a new top level folder structure called `<shared-layers>` which shall live on the same level as the `<layers>` directory.
+
+The structure would look like - 
+
+```
+<shared-layers> (configurable with `CNB_SHARED_LAYERS_DIR`)
+└── <escaped-buildpack-id> `CNB_BP_SHARED_LAYERS_DIR`
+    ├── <shared-layer>.toml
+    └── <shared-layer>
+        ├── <usable-directory>
+        └── env
+```
+
+Each buildpack will be passed an environment variable `CNB_BP_SHARED_LAYERS_DIR` with the location to the shared layers directory for a buildpack.
```

Comment:
Are we worried about collision of shared layered directories that we still want them to be "owned" by a buildpack?